sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,488 @@
+In case one `SYCL_EXTERNAL` function uses another `SYCL_EXTERNAL` function
+with different value in `sycl-module-id` attribute, the second one is not copied
+to the device image with the first function, but dependency between those device
+images is recorded instead.
+
+After device code split, all non-`SYCL_EXTERNAL` functions and copied
+`SYCL_EXTERNAL` functions left in device images with kernels marked with
+internal linkage to avoid multiple definition errors during runtime linking.
+After that `sycl-post-link` records list of names of exported functions, i.e.
+functions with `sycl-module-id` attribute and external linkage.
+
+In order to collect information about imported symbols `sycl-post-link` looks
+through LLVM IR and for each declared but not defined symbol records its
+name, except the following cases:
+- Declarations with `__` prefix in demangled name are not recorded as imported
+  functions
+  - Declarations with `__spirv_*` prefix should not be recorded as dependencies
+    since they represent SPIR-V operations and will be transformed to SPIR-V
+    instructions during LLVM->SPIR-V translation.
+- Based on some attributes (which could be defined later) we may want to avoid
+  listing some functions as imported ones
+  - This is needed to have possibility to call device-specific builtins not
+    starting with `__` by forward-declaring them in DPC++ code
+
+**NOTE**: If device code split is enabled, imports collection is performed after
+split and it is performed on separated images.
+
+All collected information is attached to a device image via properties
+mechanism.
+
+Each device image is supplied with an array of property sets:
+```
+struct pi_device_binary_struct {
+...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+
+```
+Each property set is represented by the following struct:
+```
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`SYCL/imported symbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `SYCL/exported symbols`.
+
+### DPC++ runtime changes
```

Comment:
I found it very difficult to understand the big picture for the DPC++runtime changes.  It would be nice to have an overview up front.  Maybe something like this:

> The overall strategy is as follows.  When the application submits an offload kernel to execution on a device, the runtime uses the `SYCL/imported symbols` attribute from the kernel's device image to find the set of symbols it imports.  The runtime then searches for device images that export these same symbols by searching their `SYCL/exported symbols` attributes.  This symbol resolution algorithm results in a list of device images that need to be linked together.
>
> The design currently supports only the case when all these device images are in SPIR-V format.  The runtime passes the list of device images to the PI API, which compiles and links the SPIR-V modules together into a single native device image.  This native device image is then added to the cache to avoid symbol resolution, compilation, and linking for any future attempts to invoke kernels from this device image.
>
> Note that the PI API does not use the `SYCL/imported symbols` or `SYCL/exported symbols` attributes to perform the linking of the SPIR-V modules.  Instead, functions in the SPIR-V modules must be annotated with the SPIR-V `Import` and `Export` linkage types.  The PI API layer uses these annotations to perform that actual linking of function references across SPIR-V modules.

Note that I'm not really sure this is how the design will work, but I think this is what you intend.  Can you confirm if my overall understanding of the runtime design is accurate?

Reply:
I think your understanding is correct, except that the design doesn't actually depend on device images format. I.e. the algorithm of searching won't be changed if device images are pre-compiled native device binaries.

Reply:
Sure, the symbol-search part of the algorithm can probably work for device images in native code format.  However, I think the online linking part will only work for device images in SPIR-V format.

Reply:
I also wanted to stress that I think it's important to add some description similar to what I suggest to the introduction to the "DPC++ runtime changes" section.  I think this will clarify the following points which are currently unclear:

* The `SYCL/imported symbols` and `SYCL/exported symbols` attributes are used only by the runtime's symbol resolution algorithm.  The PI API layer relies on the SPIR-V `Import` and `Export` linkage annotations, so these must also be present.

* The linking part of the design works only for device images in SPIR-V format.

* The only things added to the cache are fully linked device images.  We do not cache device images with `SYCL_EXTERNAL` functions that are compiled and not yet linked.

Reply:
> The linking part of the design works only for device images in SPIR-V format.

I don't agree with you here. Yes, right now we don't have backends that support linking of native device binaries. But we don't say that it is not possible in the future.  I think format of device image shouldn't affect design of runtime changes. Most likely when linking of native device binaries is supported, it won't matter for runtime which format device image has, it will just call some PI API for link.

Reply:
The PI API and the PI implementation layer are part of DPC++, so I think this design document should include any changes that are needed in that layer.  Currently, the PI layer assumes (I think) that it is only possible to link together SPIR-V modules.  If we want to enable the ability to link native modules, we would need some extension to the PI API.

Likewise, if the design requires changes to either the Level Zero or OpenCL interface, it should state this requirement.  Those teams won't make any changes unless we give them some requirement.

If we do not intend to do any of these things at this time, then I think it's fair to say that the design does not currently support dynamic linking of device images in native format.

At a more practical level, a reader of this document should be able to tell if there is a limitation about native format device images  That's an important limitation, which we should clearly communicate to avoid confusion later.

Reply:
> Currently, the PI layer assumes (I think) that it is only possible to link together SPIR-V modules.

I can’t actually find any specification for PI API where it says about format of input device images.
But, if this is right, we may need some additional PI API to link native device binaries and this part should be included into current design.
@kbobrovs , @smaslov-intel , could you please help to understand whether do we need some additions to PI API in order to support runtime linking of AOT-compiled binaries?

Reply:
PI defines:
```cpp
__SYCL_EXPORT pi_result
piProgramLink(pi_context context, pi_uint32 num_devices,
              const pi_device *device_list, const char *options,
              pi_uint32 num_input_programs, const pi_program *input_programs,
              void (*pfn_notify)(pi_program program, void *user_data),
              void *user_data, pi_program *ret_program);
```
this seem quite suitable to me to model linkage of programs in any state. Allowing native format for linked programs, if it is not already allowed, is not breaking backward compatibility. So I would not expect problems (unless there is some specifics in dynamic linkage not covered by this API).

Reply:
Be careful about this.  The Level Zero native code linking feature (which doesn't quite work yet), behaves quite differently from SPIR-V linking.  Native code linking does not create a new linked image.  Instead, it does in-place patching of the input images.  To illustrate, if you link native images A, B, and C together; the operation modifies all three images in-place.  (It does **not** create some new image D that represents the linked result.)  If you later link A with some other module E; this might undo the linking of A, B, and C.

There is some code in the Level Zero PI layer to handle this (which I wrote), but I'm not sure it's the right direction.  I was actually planning to remove all that code once we have SPIR-V linking in Level Zero.

Also note that this weird "in-place" patching semantic may affect the way the cache works.  Because native code linking does not create a new result image, you can't cache the resulting "linked image".

Reply:
I don't see why existing `piProgramLink` isn't sufficient. Note that it is already linking `_pi_program::Object` modules, not SPIR-V: https://github.com/intel/llvm/blob/ef4e6ddce56fe21621bca7d62932713152de3f0e/sycl/plugins/level_zero/pi_level_zero.hpp#L635

```
    // The program consists of native code (typically compiled from SPIR-v),
    // but it has unresolved external dependencies which need to be resolved
    // by linking with other Object state program(s).  Programs in this state
    // have a single Level Zero module.
```

Reply:
> Be careful about this. The Level Zero native code linking feature (which doesn't quite work yet), behaves quite differently from SPIR-V linking. Native code linking does not create a new linked image. Instead, it does in-place patching of the input images.

Understood, thanks. Still, I don't see how this affects the PI API? It affects implementation - true. LevelZero plugin might need to copy before calling LevelZero linkage. - ?

But in-place patching is really not quite intuitive - I agree.

> Also note that this weird "in-place" patching semantic may affect the way the cache works. Because native code linking does not create a new result image, you can't cache the resulting "linked image".

Why would we need to cache anything when the code is already compiled (native)?

Reply:
> Why would we need to cache anything when the code is already compiled (native)?

My understanding of the design is that with SPIR-V modules we will link all the dependent modules together first and then cache the result.  My point is that the strategy will be different for native modules.  It does not make sense to cache the result of native linking because of the in-place semantic.  Therefore the caching strategy is different between SPIR-V and native cases.

> Understood, thanks. Still, I don't see how this affects the PI API? It affects implementation - true. LevelZero
> plugin might need to copy before calling LevelZero linkage. - ?

Perhaps we will decide to use the same PI API in both cases.  My point is that something here will need to change.  Either we add a new PI API or we re-use the existing API for both SPIR-V and native cases.  Either way, something needs to change.  The main goal of my comment here was to point out that this area of the runtime will need changes if we want to support native modules, and there is no mention of that in the design document.

Reply:
> It does not make sense to cache the result of native linking because of the in-place semantic.

I think this in-place semantic is implementation detail of PI Level Zero plugin, because it might not be the case for OpenCL, for example. And we are actually in control of the API here: we can say that `piProgramLink` must not modify inputs and therefore to DPC++ layer this in-place semantic wouldn't matter at all. It would just be hidden somewhere in implementation below.

Reply:
> this seem quite suitable to me to model linkage of programs in any state. Allowing native format for linked programs, if it is not already allowed, is not breaking backward compatibility. So I would not expect problems (unless there is some specifics in dynamic linkage not covered by this API).

I think one more important thing here is that DPC++ RT should be able to understand whether it can pass native binaries to link API or not: at the moment we don't have capabilities to link native binaries with all underlying runtimes we support so that is important. I was thinking about new PI interface or extension to `piDeviceGetInfo` - @smaslov-intel, @kbobrovs what do you think?

Reply:
@AlexeySachkov , extending the `piDeviceGetInfo` would be reasonable if really needed. What are the backends that cannot link native binaries? From https://www.khronos.org/registry/OpenCL//sdk/2.2/docs/man/html/clLinkProgram.html:

> input_programs is an array of program objects that are compiled binaries or libraries that are to be linked to create the program executable. 

CUDA also takes PTX as an input: https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MODULE.html#group__CUDA__MODULE_1g3ebcd2ccb772ba9c120937a2d2831b77

Reply:
Extending piDeviceGetInfo makes sense to me.

One can say that SDV who uses DPC++ should know if particular device supports native linkage or not and prepare binaries accordingly, and that would be true. But in presence of both forms - SPIRV and native (maybe more in theory), of the dynamically linked app and device libraries, this can be a decisive factor for SYCL RT. And the extension of the GetInfo is really cheap.

> What are the backends that cannot link native binaries?

I believe Intel OpenCL GPU is one example. There can be more from other vendors.

Reply:
> I believe Intel OpenCL GPU is one example.

Isn't the excerpt from OpenCL API I gave above telling us it is linking "native binaries"?

Reply:
> Isn't the excerpt from OpenCL API I gave above telling us it is linking "native binaries"?

You should check with the OpenCL team.  My understanding is that `clCompileProgram()` is a no-op and `clLinkProgram()` actually does a combined SPIR-V link + compile operation.  I do not think it is possible to use `clLinkProgram()` to link native modules together if you no longer have the original SPIR-V.

Reply:
@smaslov-intel, my understanding is that in OpenCL terminology, `compiled` != `built`. `compiled` is an object form, i.e. front-end is passed, but no executable binary was produced yet. I can't find strong proof of that in the spec, but I'm sure that our (Intel) CPU & GPU compilers won't accept native binaries in `clLinkProgram`

Reply:
https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clCompileProgram.html

> The compiled binary can be queried using clGetProgramInfo(program, CL_PROGRAM_BINARIES, ...) and can be specified to clCreateProgramWithBinary to create a new program object.

Compile does produce "binary", otherwise the above wouldn't be possible.
Who can clarify it exactly?

Reply:
> Compile does produce "binary", otherwise the above wouldn't be possible.
> Who can clarify it exactly?

I think @bashbaug should be able to clarify this.

In the meantime: 


From [clCreateKernel](https://www.khronos.org/registry/OpenCL/sdk/2.0/docs/man/xhtml/clCreateKernel.html):
> Parameters
> program
> A program object with a successfully built executable.

From [clBuildProgram](https://www.khronos.org/registry/OpenCL/sdk/2.0/docs/man/xhtml/clBuildProgram.html):
> Builds (compiles and links) a program executable from the program source or binary.

From [clLinkProgram](https://www.khronos.org/registry/OpenCL/sdk/2.0/docs/man/xhtml/clLinkProgram.html):
> Links a set of compiled program objects and libraries for all the devices or a specific device(s) in the OpenCL context and creates an executable.

From [clCompileProgram](https://www.khronos.org/registry/OpenCL/sdk/2.0/docs/man/xhtml/clCompileProgram.html):
> Compiles a program’s source for all the devices or a specific device(s) in the OpenCL context associated with program.

It seems to me that intent was that `clCompileProgram` does not generate an executable, but only creates some intermediate object file instead

Reply:
+1.
From clGetProgramInfo(https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clGetProgramInfo.html) desc:

> Return the program binaries for all devices associated with program. For each device in program, the binary returned can be the binary specified for the device when program is created with clCreateProgramWithBinary or it can be the executable binary generated by clBuildProgram. If program is created with clCreateProgramWithSource, the binary returned is the binary generated by clBuildProgram. The bits returned can be an implementation-specific intermediate representation (a.k.a. IR) or device specific executable bits or both. The decision on which information is returned in the binary is up to the OpenCL implementation.

Reply:
> The decision on which information is returned in the binary is up to the OpenCL implementation.

Can we somehow query OpenCL about what is in the binary such that we could properly control it's linking?